### PR TITLE
Add sf::PrimitiveInfo

### PR DESCRIPTION
**Pull request by Marioalexsan**
## Description

This PR adds the `sf::PrimitiveInfo` interface, which allows extracting vertex info from drawables. The interface is implemented for every drawable except `sf::VertexBatch`. This allows users to use SFML's drawables together with custom render logic (such as sprite batching), instead of recreating all of the logic using `sf::VertexArray` in specialized classes.

## Tasks

* [ ] Tested on Linux
* [X] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

Render some drawables using this code. You will need a TTF and a PNG:
```cpp
#include <SFML/Graphics.hpp>

int main()
{
    sf::RenderWindow window(sf::VideoMode(1280, 720), "PrimitiveInfo", sf::Style::Default);
    window.setView({ { 0, 0 }, { 1280, 720 } });

    sf::Font font;
    sf::Texture texture;

    if (!font.loadFromFile("Font0.ttf") || !texture.loadFromFile("Texture0.png"))
    {
        sf::err() << "Oh no!";
        return 0;
    }

    sf::Text text("Primitives are cool", font, 30);
    text.setPosition({ 0, 0 });

    sf::Sprite sprite(texture);
    sprite.setPosition({ 200, 0 });

    sf::CircleShape shape(100, 6);
    shape.setPosition({ 400, 0 });
    shape.setTexture(&texture);

    const sf::Texture* drawTexture;
    sf::Transform transform;

    auto renderer = [&text, &window, &drawTexture, &transform](std::vector<sf::Vertex>& vertices, sf::PrimitiveType type)
    {
        for (auto& vertex : vertices)
        {
            vertex.position = transform.transformPoint(vertex.position);
        }

        window.draw(vertices.data(), vertices.size(), type, sf::RenderStates(drawTexture));
    };

    drawTexture = &text.getFont()->getTexture(text.getCharacterSize());
    transform = text.getTransform();

    text.forEachMutablePrimitive(renderer);

    drawTexture = sprite.getTexture();
    transform = sprite.getTransform();

    sprite.forEachMutablePrimitive(renderer);

    drawTexture = shape.getTexture();
    transform = shape.getTransform();

    shape.forEachMutablePrimitive(renderer);

    window.display();

    sf::sleep(sf::seconds(6));
}
```

### CODE REVIEW

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by Bromeon** [comment]
Thanks for your contribution!

This seems to be a highly specialized API, and the callback-based approach is rather unconventional for SFML's style. Also, transform being outside the callback is probably not that useful, as transforms (unlike textures) are rarely reused across a large number of `sf::Drawable` objects (except in `sf::VertexArray`).

What is the concrete use case you have in mind? Are you using this pattern in an application?
The example code you show is mostly `window.draw()` with extra steps 🙂

**Comment by Marioalexsan** [comment]
The test code is lame indeed... A better use for this API is to implement sprite batches. The main benefit is that one can use SFML's drawables (or custom drawables) directly, instead of reimplementing `sf::Sprite` or using `sf::VertexArray` with custom logic.

The concrete use case for this API would be to implement sprite batches. I have created this [repo](https://github.com/Marioalexsan/SFMLBatchTest/blob/master/GraphicsBatch.cpp) that uses the PrimitiveInfo API to batch sprites and other things.

This [repo](https://github.com/ricanteja/Moony-SpriteBatch) is an example of where sprite batching is implemented

**Comment by Marioalexsan** [comment]
The main use case of this API would be to implement sprite batches that can work directly with SFML's drawables. Currently, users have to roll their own logic using `sf::VertexArray` and reimplement `sf::Sprite` and `sf::Text` to be batchable. For example, [this repository](https://github.com/ricanteja/Moony-SpriteBatch/tree/master/source) uses a custom (and probably incomplete) Sprite class. #1802 uses `sf::VertexArray`, which would also require users to reimplement drawables.

While I don't have any cool production code or anything, I have [this sprite batch class](https://github.com/Marioalexsan/SFMLBatchTest/blob/master/GraphicsBatch.cpp) that utilizes the API. While the test case is lame (since it's just a fancy `window.draw()` like you mentioned), the batch class defers rendering until `batch.end()` is called, where it can reorder sprites and render them with less render state switching.

File: include/SFML/Graphics/PrimitiveInfo.hpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    /// \brief Virtual destructor
+    ///
+    ////////////////////////////////////////////////////////////
+    virtual ~PrimitiveInfo() {}
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Callback that uses constant vertex info
+    ///
+    ////////////////////////////////////////////////////////////
+    typedef std::function<void(const Vertex* vertices, std::size_t vertexCount, PrimitiveType type)> PrimitiveCallback;
```
Comment (vittorioromeo) [inline include/SFML/Graphics/PrimitiveInfo.hpp:59]:
```suggestion
    using PrimitiveCallback = std::function<void(const Vertex* vertices, std::size_t vertexCount, PrimitiveType type)>;
```

File: include/SFML/Graphics/PrimitiveInfo.hpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Callback that uses mutable vertex info
+    ///
+    ////////////////////////////////////////////////////////////
+    typedef std::function<void(std::vector<Vertex>& vertices, PrimitiveType type)> MutablePrimitiveCallback;
```
Comment (vittorioromeo) [inline include/SFML/Graphics/PrimitiveInfo.hpp:65]:
```suggestion
    using MutablePrimitiveCallback = std::function<void(std::vector<Vertex>& vertices, PrimitiveType type)>;
```

File: include/SFML/Graphics/PrimitiveInfo.hpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    /// the derived class to define what primitive info is being
+    /// provided.
+    ///
+    /// While the order of the primitives is not specified, it
+    /// would be a good idea to iterate them in draw order.
+    ///
+    /// \param callback The function to call for each primitive
+    ///
+    ////////////////////////////////////////////////////////////
+    virtual void forEachPrimitive(PrimitiveCallback callback) const = 0;
```
Comment (vittorioromeo) [inline include/SFML/Graphics/PrimitiveInfo.hpp:80]:
```suggestion
    virtual void forEachPrimitive(const PrimitiveCallback& callback) const = 0;
```

File: include/SFML/Graphics/PrimitiveInfo.hpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    /// \brief Iterate over each primitive and call the function
+    ///
+    /// This is a helper method that calls the virtual overload,
+    /// and puts the results inside a mutable std::vector<Vertex>
+    /// parameter.
+    ///
+    /// \param callback The function to call for each primitive
+    ///
+    ////////////////////////////////////////////////////////////
+    void forEachMutablePrimitive(MutablePrimitiveCallback callback) const;
```
Comment (vittorioromeo) [inline include/SFML/Graphics/PrimitiveInfo.hpp:92]:
```suggestion
    void forEachMutablePrimitive(const MutablePrimitiveCallback& callback) const;
```

File: src/SFML/Graphics/PrimitiveInfo.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+{
+    std::vector<Vertex> vec;
+
+    forEachPrimitive([&](const Vertex* vertices, std::size_t vertexCount, PrimitiveType type) {
+        vec.resize(vertexCount);
+
+        for (std::size_t i = 0; i < vertexCount; i++)
+        {
+            vec[i] = vertices[i];
+        }
```
Comment (vittorioromeo) [inline src/SFML/Graphics/PrimitiveInfo.cpp:45]:
```suggestion
        vec.assign(vertices, vertices + vertexCount);
```

**Review by vittorioromeo — CHANGES_REQUESTED** [review]
C++-related changes. Have not thought about the use cases for this feature yet.

**Comment by Marioalexsan** [comment]
I'm not really satisfied with my approach, so I'll clean this up and open another pull request when ready.

**Comment by Bromeon** [comment]
Maybe it makes sense to discuss the design first (in an issue or on the forum). Not that you put all the effort into a PR and then have to throw it away.

**Pull request closed without merging**